notebooks/time_series/apple_stock_prices.livemd

```diff
@@ -0,0 +1,264 @@
+<!-- livebook:{"file_entries":[{"name":"Apple_Stock_Prices_1981_to_2023.csv","type":"attachment"},{"name":"apple_prices.png","type":"attachment"},{"name":"normalized_apple_prices.csv","type":"attachment"},{"name":"predictions.png","type":"attachment"},{"name":"train_test_plt.png","type":"attachment"}]} -->
+
+# Time-series regression with RNNs
+
+```elixir
+Mix.install([
+  {:axon, "~> 0.6.1"},
+  {:kino_vega_lite, "~> 0.1.13"},
+  {:nx, "~> 0.7.3"},
+  {:tucan, "~> 0.3.1"},
+  {:polaris, "~> 0.1.0"},
+  {:exla, "~> 0.7.3"},
+  {:req, "~> 0.5.2"},
+  {:vega_lite, "~> 0.1.9"}
+])
+
+Nx.global_default_backend(EXLA.Backend)
+Nx.Defn.global_default_options(compiler: EXLA)
+```
+
+## Introduction
+
+In this project, we will build a predictor for a time series using an Recurrent Neural Network regressor. The prediction will be made for Apple's stock price 7 days in advance, based on a historical series.
+
+An architecture known as Long Term Short Memory (LSTM) will be used for the RNN.
+
+<!-- livebook:{"break_markdown":true} -->
+
+First, we load the historical data. Subsequently, preprocessing will be conducted to utilize the information with an RNN model. The time series data is already normalized. This helps mitigate significant numerical issues associated with how activation functions like tanh transform very large numbers (whether positive or negative), and it aids in avoiding problems with calculating derivatives.
+
+```elixir
+raw_csv =
+  "https://raw.githubusercontent.com/santiago-imelio/datasets/main/normalized_apple_prices.csv"
```

Comment:
Hmm, perhaps we look for another place to host this

Reply:
Sure, where do you suggest to upload it? I can put it on Kaggle but I won't be able to download it using `Req`.

Reply:
We will leave it for now :)

Reply:
Oh, I just realized this data is already normalized. Perhaps you can use the unnormalized prices instead, and then show how to do normalization on the data with Explorer, and then converting to tensor? This would probably be a more common workflow anyway. What do you think?

Reply:
Hey, I updated the notebook to use an unnormalized dataset of stock prices. The data used is an extract of [this Kaggle dataset](https://www.kaggle.com/datasets/tarunpaparaju/apple-aapl-historical-stock-data). Then I added a normalization section using Scholar.